website/source/docs/connect/proxies/envoy.md

```diff
@@ -14,257 +14,309 @@ optionally exposing a gRPC service on the local agent that serves [Envoy's xDS
+of a [proxy service
+definition](/docs/connect/proxies.html#proxy-service-definitions) or
+[`sidecar_service`](/docs/connect/proxies/sidecar-service.html) block.
+
+- `envoy_statsd_url` - A URL in the form `udp://ip:port` identifying a UDP
+  statsd listener that metrics should be delivered to. For example this may be
+  `udp://127.0.0.1:8125` if every host has a local statsd listener. In this
+  case it is convenient to configure this property once in the [global proxy
+  defaults](#TODO) config entry. Currently, TCP is not supported.
+
+    -> **Note:** currently the url **must use an ip address** not a dns name due
+    to the way Envoy is setup for statsd.
+
+    The whole parameter may also be specified in the form `$ENV_VAR_NAME` which
+    will cause the `consul connect envoy` command to resolve the actual URL from
+    the named environment variable when it runs. This for example allows each
+    pod in a Kubernetes cluster to learn of a pod-specific IP address for statsd
+    when the envoy instance is bootstrapped while still allowing global
+    configuration of all proxies to use statsd in the [global `proxy-defaults`
+    configuration entry](#TODO). The env variable must contain a full valid URL
+    value as specified above and nothing else. It is not possible to use
+    environment variables as only part of the URL currently.
+
+- `envoy_dogstatsd_url` - The same as `envoy_statsd_url` with the following
+  differences in behavior:
+    - Envoy will use dogstatsd tags instead of statsd dot-separated metric names
+    - As well as `udp://`, a `unix://` URL may be specified if your agent can
+      listen on a unix socket (e.g. the dogstatsd agent).
+
+- `envoy_prometheus_bind_addr` - Specifies that the proxy should expose a Prometheus
+  metrics endpoint to the _public_ network. It must be supplied in the form
+  `ip:port` and port and the ip/port combination must be free within the network
+  namespace the proxy runs. Typically the IP would be `0.0.0.0` to bind to all
+  available interfaces or a pod IP address.
+
+    -> **Note:** Envoy versions prior to 1.10 do not export timing histograms
+    using the internal prometheus endpoint. Consul 1.5.0 [doesn't yet support
+    Envoy 1.10](#supported-versions) although support will soon be added.
+
+- `envoy_stats_tags` - Specifies one or more static tags that will be added to
+  all metrics produced by the proxy.
+
+- `envoy_stats_flush_interval` - Configures Envoy's
+  [`stats_flush_interval`](https://www.envoyproxy.io/docs/envoy/v1.9.1/api-v2/config/bootstrap/v2/bootstrap.proto#envoy-api-field-config-bootstrap-v2-bootstrap-stats-flush-interval).
+
+There are more possibilities available in the [Advanced
+Configuration](#advanced-configuration) section that allow incremental or
+complete control over the bootstrap configuration generated.
+
+## Dynamic Configuration
+
+Envoy's dynamic configuration is generated by Consul and is automatic. For Envoy
```

Comment:
```suggestion
Consul automatically generates Envoy's dynamic configuration. For Envoy
```

Reply:
I'm not exactly sure what is happening here, and in the next paragraph.
- Are we saying that though the config generation is automatic it's influenced by the service defaults?
- How is this connected to the proxy defaults mentioned in the next paragraph?

Reply:
Good questions. The answer is that we really want to push people towards specifying things like `protocol` for services centrally via `service-defaults` but in practice all that does is automatically add the config into the proxy config or upstream config.

The challenge here is documenting the config keys available in a proxy service's `Proxy.Config` and `Proxy.Upstreams[*].Config` while trying to encourage the use of centralised config to manage them rather than hard-coding these into registrations.

> Are we saying that though the config generation is automatic it's influenced by the service defaults?

Yes. The automatic part might be unnecessary but was trying to set the tone for why there are relatively few knobs here - we provided an opinionated config based on protocol. Later there will be more possible here but based on higher-level things like routing config rather than specific knobs for Envoy etc.

> How is this connected to the proxy defaults mentioned in the next paragraph?
The `protocol` is actually passed to the Envoy in the `Proxy.Config` which is why that needs to be documented.  We want to encourage users not to hard-code that in their own registrations though (because then it doesn't get automatically discovered by other services using their service as an upstream) and use `service-defaults` instead. When `service-defaults` is used, we still actually pass the config to Envoy via the `Proxy.Config` map in the same way as if user set them manually, we just automatically add those and keep them in sync for all service instances etc.

If you can think of a clearer way to explain this somewhere in this document or elsewhere I'm all ears. I'm not really happy with how it ended up but just wanted something reasonably complete to get feedback on.